django/core/checks/urls.py

```diff
@@ -64,16 +64,38 @@ def check_pattern_startswith_slash(pattern):
     """
     Check that the pattern does not begin with a forward slash.
     """
-    regex_pattern = pattern.regex.pattern
-    if regex_pattern.startswith('/') or regex_pattern.startswith('^/'):
+    errors = []
+
+    if not hasattr(pattern, 'regex'):
         warning = Warning(
-            "Your URL pattern {} has a regex beginning with a '/'. "
-            "Remove this slash as it is unnecessary.".format(describe_pattern(pattern)),
-            id="urls.W002",
+            "Your URL pattern {} is not an object returned from the 'url()'"
+            "method. If you are using a custom object, it should provide a `pattern`"
```

Comment:
I'm a bit hesitant about exposing the private details of url resolving in user-facing messages, so I would remove the comment about a custom object. Custom url objects are an advanced and unsupported use-case (for now), and I think this warning should target the mistakes made when upgrading/using the new structure, not the ones made when replacing the internal and private details of url resolving.